### PR TITLE
End to End testing

### PR DESCRIPTION
**Pull request by JulienBalestra**
### What does this PR do?

An introduction of the end to end testing pipeline.

Docs on a separate PR.

### CODE REVIEW

**Review by mfpierre — CHANGES_REQUESTED** [review]
A few nits on bash scripts & agent memory limits

File: test/e2e/argo-workflows/agent.yaml [hunk 1]
```diff
@@ -0,0 +1,537 @@
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: argo-datadog-agent-
+spec:
+  entrypoint: main
+  #onExit: delete # call argo submit --entrypoint delete instead
```
[the remark this answers is not in the document]
Comment (JulienBalestra) [reply test/e2e/argo-workflows/agent.yaml]: No, this an helper to help you to clean your Kubernetes cluster.

[the remark this answers is not in the document]
Comment (JulienBalestra) [reply test/e2e/argo-workflows/agent.yaml]: An another use-case could be: onExit: upload all the test-data to a dedicated bucket.

**Review by hkaj — COMMENTED** [review]
fake_datadog looks good to me

File: test/e2e/scripts/setup-instance/02-ec2.sh [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    --valid-until $(($(date +%s) + 3300)) \
+    --launch-specification file://specification.json \
+    --region ${REGION} | tee spot-instance-request.json
+
+SPOT_REQUEST_ID=$(jq -re .SpotInstanceRequests[].SpotInstanceRequestId spot-instance-request.json)
+
+
+while true
+do
+    aws ec2 describe-spot-instance-requests \
```
Comment (CharlyF) [inline test/e2e/scripts/setup-instance/02-ec2.sh]: Depending on how the aws cli is installed and the creds are set, would it make sense to make sure we can request/create an instance before looping here ?

File: test/e2e/scripts/setup-instance/03-ssh.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+until _ssh /bin/true
+do
+    sleep 5
+done
+
+
+_ssh git clone https://github.com/DataDog/datadog-agent.git /home/core/datadog-agent || {
+    # To be able to retry
+    echo "Already cloned ?"
```
Comment (CharlyF) [inline test/e2e/scripts/setup-instance/03-ssh.sh]: In which case would the repo be already cloned ?

**Review by CharlyF — COMMENTED** [review]
That looks good, I left a few questions.

File: test/e2e/scripts/setup-instance/01-ignition.sh [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        "contents": {
+          "source": "data:,GROUP%3Dalpha%0AREBOOT_STRATEGY%3Doff%0A"
+        },
+        "filesystem": "root"
+      },
+      {
+        "path": "/opt/bin/setup-pupernetes",
+        "mode": 320,
+        "contents": {
+          "source": "data:,%23%21%2Fbin%2Fbash%20-ex%0Acurl%20-Lf%20https%3A%2F%2Fgithub.com%2FDataDog%2Fpupernetes%2Freleases%2Fdownload%2Fv0.2.1%2Fpupernetes%20-o%20%2Fopt%2Fbin%2Fpupernetes%0Asha512sum%20-c%20%2Fopt%2Fbin%2Fpupernetes.sha512sum%0Achmod%20%2Bx%20%2Fopt%2Fbin%2Fpupernetes%0A"
```
[the remark this answers is not in the document]
Comment (JulienBalestra) [reply test/e2e/scripts/setup-instance/01-ignition.sh]: No, it should be documented in the next docs PR 👍 
It's available here right now: [release process](https://github.com/DataDog/pupernetes/blob/master/RELEASE.md)

File: test/e2e/scripts/setup-instance/02-ec2.sh [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+COMMIT_USER=$(git log -1 --pretty=format:'%an')
+
+# If not using the default value, mind to change the following settings:
+# - AMI
+# - security groups ids
+# - subnet ids
+REGION="${REGION:-us-east-1}"
+
+aws ec2 request-spot-instances \
+    --spot-price "0.015" \
```
[the remark this answers is not in the document]
Comment (JulienBalestra) [reply test/e2e/scripts/setup-instance/02-ec2.sh]: There is some apps online to observe the spot price, I took something higher to be confortable.
Do you have any advice for that ?

File: test/e2e/scripts/setup-instance/03-ssh.sh (hunk 3 above)
Comment (JulienBalestra) [reply test/e2e/scripts/setup-instance/03-ssh.sh]: When we are iterating on the dev target. The instance could be partially already prepared.

This practice have a lot of limitations, but accept an already cloned repo sounds reasonable to me.

File: test/e2e/scripts/setup-instance/02-ec2.sh (hunk 2 above)
Comment (JulienBalestra) [reply test/e2e/scripts/setup-instance/02-ec2.sh]: Why not, how could we do that ?

File: test/e2e/scripts/setup-instance/02-ec2.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (CharlyF) [reply test/e2e/scripts/setup-instance/02-ec2.sh]: Just curious, making sure we don't spend too much on the e2e tests.
If you use a legit source it's all good though

File: test/e2e/scripts/setup-instance/03-ssh.sh (hunk 3 above)
Comment (CharlyF) [reply test/e2e/scripts/setup-instance/03-ssh.sh]: I see, I don't have much experience there but accepting an already cloned repo means potentially being out of date so I think this is legit.

File: test/e2e/scripts/setup-instance/02-ec2.sh (hunk 2 above)
Comment (CharlyF) [reply test/e2e/scripts/setup-instance/02-ec2.sh]: Something along the lines of https://docs.aws.amazon.com/IAM/latest/APIReference/API_GetRole.html ?

File: test/e2e/scripts/setup-instance/03-ssh.sh (hunk 3 above)
Comment (JulienBalestra) [reply test/e2e/scripts/setup-instance/03-ssh.sh]: I encountered this issue while developing on this.

This mean, if you did a typo in any script, without this, you need to create a new instance to try the fix.

File: test/e2e/scripts/setup-instance/02-ec2.sh (hunk 2 above)
Comment (JulienBalestra) [reply test/e2e/scripts/setup-instance/02-ec2.sh]: We are already running with a custom IAM roles, all the permissions are defined before and on a separate repo.

Comment (CharlyF) [reply test/e2e/scripts/setup-instance/02-ec2.sh]: yea, I understand. I just figured that if someone modifies the perms, it might affect the e2e

**Pull request merged**